### ISSUE TITLE
PHP 7

### ISSUE THREAD

**Issue by gburton** [gburton/Responsive-osCommerce#288]
**PHP7 updates**

Methods with the same name as their class will not be constructors in a future version of PHP in module files and class files

**Admin Side**
- [x] classes - assigned to @BrockleyJohn 
- [ ] functions
- [ ] modules
- [ ] other

**Shop Side**
- [ ] classes
- [ ] functions
- [ ] modules/action_recorder
- [ ] modules/boxes
- [ ] modules/content
- [ ] modules/header_tags
- [ ] modules/navbar_modules
- [ ] modules/order_total
- [ ] modules/payment
- [ ] modules/shipping
- [ ] modules/social_bookmarks
- [ ] other

**Comment by Piernas** [comment]
Do you want them to be done all at once?

**Comment by gburton** [comment]
@Piernas no.  can be done bit by bit (and bit by bit is easier for testing, pushing, and committing).

**Comment by Piernas** [comment]
Ok, I'll do class constructors as I found them.

**Comment by gburton** [comment]
That would be difficult.  Best would be to put your name against one lot of files (eg) /includes/modules/header_tags/ so that others know you are doing those.  And someone else does not then also do it.  Does it make sense ?

**Comment by Piernas** [comment]
Perfect. but I'm seeing it's not much work. I can do all the modules in an hour.

**Comment by gburton** [comment]
@piernas **Please ** do only one lot of modules as I suggested.  Then test them.  Then push...

**Comment by BrockleyJohn** [comment]
ok - put me down for admin/classes
...testing might take a while!

Another php7 pitfall is that functions called statically must  be so declared but I think the only bit of core with any static anything is Braintree payments

If we're doing piecemeal, we'll not be able to test on php7 until a coherent chunk is done. I assume we're just renaming the constructor to __construct (and leaving the documenting to github) and testing that the constructor still gets executed and performs as expected. Will keep a note of what I did for each test.

**Comment by gburton** [comment]
@BrockleyJohn exactly that, and thank you.  admin/classes are yours.

**Comment by BrockleyJohn** [comment]
Actually this approach isn't sufficient; it doesn't cater for the constructor method being called directly as I have found in testing. A general approach to keep changes and testing to a minimum is to create a constructor __construct that calls the existing class name function.

**Comment by BrockleyJohn** [comment]
further - and probably irrelevant - thought:

__construct calling the class name constructor for all classes would preserve backward compatibility to php4
!!

**Comment by BrockleyJohn** [comment]
and another, what about
`$obj = new $class($params);
...
$result = $obj->$class($newparams);`
maybe we should always be defining the new constructor to call the existing one

**Comment by BrockleyJohn** [comment]
@gburton you can put me down for admin / modules too now...

**Comment by BrockleyJohn** [comment]
@gburton in the latest set of updates I hit another php7 incompatibility relating to the interpretation order in indirect variable handling (discussed [http://forums.oscommerce.com/topic/409316-php7-dynamic-classarray-handling-gotcha/?p=1742119](url)). I'm going to have a search through the codebase and see if I can spot any other occurrences

**Comment by BrockleyJohn** [comment]
@gburton here's another incompatibility:

> global only accepts simple variables 
> Variable variables can no longer be used with the global keyword. The curly brace syntax can be used to emulate the previous behaviour if required

so in functions/database.php
`global $$link`
needs to become
`global ${$link}`
there may be other cases...

**Comment by BrockleyJohn** [comment]
put me down for admin / functions next
... not much to change, I hope!

**Comment by BrockleyJohn** [comment]
@gburton I've checked through the rest of admin and not found any more changes to make. There are still warnings in my test admin environment but they all come from files on the catalog side.

I'll start on the catalog modules next: 
action_recorder is a nice short list

**Comment by raiwa** [comment]
I have prepared the following on shop side:
classes
modules/action_recorder
modules/boxes
modules/header_tags
modules/order_totals
modules/social_bookmarks

If assigned I could upload them within a couple of days

**Comment by BrockleyJohn** [comment]
@raiwa Rainer, pick one of the six for the first to put in a branch, double-check the testing then create a pull request and it's yours - perhaps classes?
Then on to whichever you want next but let us know which you're doing, please.

It's self-assignment, but one unit at a time!

In view of your list I'll pick up the payment modules next despite expecting proper testing to be a nightmare ...

**Comment by raiwa** [comment]
OK, I'll do classes tomorrow.

**Comment by MOPS-DanCole** [comment]
Rainer can you add our shopping cart class to your list...apparently it's tossing errors or at least warnings anyway.

**Comment by raiwa** [comment]
@BrockleyJohn, forked, checked, placed pull request #335.
How do I self assign the Shop classes?

**Comment by raiwa** [comment]
Hello Dan,

 

The shopping cart class PHP7 update is already included in my PHP7 update list.

 

I have separated the checkout modules and send you attached the 2 new modules.

I added:

 

                    if ($cart->count_contents() > 0) {

 

checks to both modules.

 

For the alternative checkout module I added a check for the order of both modules to show the “or” text in the correct place, or not to show it if the other module is not active.

However if they are shown one beside the other it will not work correct, but we can’t take care for all posible settings.

Or any idea for that case?

 

Kind regards

Rainer

 

De: Dan Cole [mailto:notifications@github.com] 
Enviado el: martes, 12 de abril de 2016 19:48
Para: gburton/Responsive-osCommerce
CC: Rainer Schmied
Asunto: Re: [gburton/Responsive-osCommerce] PHP 7 (#288)

 

Rainer can you add our shopping cart class to your list...apparently it's tossing errors or at least warnings anyway.

**Comment by BrockleyJohn** [comment]
@raiwa - you already assigned yourself the shop classes:
> OK, I'll do classes tomorrow.

That's all there is to it! Gary's too busy to keep updating the first post every five minutes! When you're ready to do another one, just post it here.

**Comment by BrockleyJohn** [comment]
@raiwa Rainer, you said before you've done some of the modules as well as the classes. Do you still have that code and if so, would you put it into a branch for each module type and push, please?

I'm going to start on modules/shipping next

**Comment by raiwa** [comment]
@BrockleyJohn, ok I'll do action recorder then

**Comment by raiwa** [comment]
@BrockleyJohn I did action recorder and also boxes

**Comment by boelle** [comment]
Side Q... will there be an manual upgrade guide for those of us that have modified the shop?

unoeuro have allready started to roll out PHP 7

**Comment by boelle** [comment]
or will we be on our own?

**Comment by MOPS-DanCole** [comment]
boelle...my guess is that you'll be on your own.....osC is trying to discourage users from making changes to the core so it's unlikely that anyone is going to want to spend time/resources in that way.

**Comment by boelle** [comment]
does that also count for the responsice osc ?

all i'm after is what i need to do to 2.3.4 responsive to be able to run on  php 7

**Comment by MOPS-DanCole** [comment]
Edge is being updated to run on PHP 7...when it's ready you'll need to either...

1. install it and add your modifications OR
2. use a compare program and add the changes that were necessary to your modified shop to get it to work up php7.

**Comment by BrockleyJohn** [comment]
The php7 changes aren't very complicated - and aren't very likely to break customisation as long as you know which files you've changed. If you want to apply the changes to your own shop code before they've been fully tested and released that is an option for you - check out the pull requests for the completed work. I'd be surprised if hosts are forcing people onto php7 yet, surely it's just being rolled out as an option? 

Anyway, the manual upgrade guide boils down to:
rename the class constructor in all classes 
from function classname($params) 
to function __construct($params)
and check there aren't any calls to $obj->classname

**Comment by boelle** [comment]
Nice

and i guess i can limit this work to the list of files Gary has put in the first post?

**Comment by BrockleyJohn** [comment]
if you note that there's a point 'other' in Gary's list, you'll realise it applies to the whole of osc (including /ext)

**Comment by BrockleyJohn** [comment]
@boelle if you're keen to have a go, why not start with 
modules/social boookmarks
and when you've tested them, create a pull request?

**Comment by boelle** [comment]
actually i'm trying to get edge up on a web space that runs php7.

it will be both a live shop and a playground of sorts.... currently stuck in the dreaded redirect loop at admin

**Comment by boelle** [comment]
looking at the sb_digg module, if i understand right line 38

function sb_digg()

is renamed to  

function __construct()

?

**Comment by BrockleyJohn** [comment]
Yes on the change, but I'd expect you to find it on line 21. 

BTW there's a full set of all the admin changes in the pull requests - it's the shop side that's unfinished.

**Comment by boelle** [comment]
i saw those...  planning to implement them also

I do have the option to downgrade php via htaccess so i can jump back and forth 

also agree on line 21... not sure where i got 38 from

last Q this time round... how do i test individual files ? of course set for php7 but then what?

**Comment by boelle** [comment]
btw... fork is here: https://github.com/boelle/Responsive-osCommerce

all the social bookmarks are done, only need to test them before i do a PR

**Comment by boelle** [comment]
i can take shipping modules too

for flat rate i guess line 17 is the one

**Comment by boelle** [comment]
i assumed it was right and now done with the shipping folder

**Comment by BrockleyJohn** [comment]
You need to do one set of changes on their own in their own branch from master which must be in step with Gary's and then raise a pull request

**Comment by boelle** [comment]
bummer....

is all work lost or can i save it somehow? not a github geek by any means, just have time to spend doing this simple text edit

**Comment by boelle** [comment]
think i managed to do a branch that holds shipping and social bookmarks... 

created a PR for it

**Comment by BrockleyJohn** [comment]
No just one for social bookmarks. I've already done shipping

**Comment by boelle** [comment]
upps... will try and remove the shipping ones then

how do i test my work and what could i attack next?

**Comment by BrockleyJohn** [comment]
You shouldn't raise a PR until you've tested it! 
Please kill it until you have.
For testing you should:
- check that unchanged module produces a deprecated notice
- check that modified module does not
- confirm that module installs correctly
- confirm that module executes correctly
(the last two are a simple way of checking for code outside the module that relies on the old classname function)

**Comment by boelle** [comment]
done :-D

what could i have a go at next?

**Comment by BrockleyJohn** [comment]
how about modules / content? that should keep you busy ;)

**Comment by boelle** [comment]
yep

@gburton can you update the list so everyone has an quick list of who is doing what

**Comment by BrockleyJohn** [comment]
still to do are (shop side):
 functions
 modules/header_tags
 modules/navbar_modules
 modules/order_total
 other

**Comment by boelle** [comment]
might pick one of those instead but will give a shout before i do

in the process of installing all the PR's that relates to php7. So far i have not encountered any problem

but a bit tired but tomorrow i will use a compare tool and do all the classes

**Comment by boelle** [comment]
will do these 2 during the day if no one is on them

modules/navbar_modules
modules/order_total

**Comment by boelle** [comment]
hmm... just did navbar modules but i cant see them anywhere in admin... am i looking the wrong place maybe?

**Comment by boelle** [comment]
anyways both folders should be ok

modules/header_tags will be next when i get the nav bar modules to show up

**Comment by boelle** [comment]
figured the nav bar thing... i  was missing a lot of files since i dont use this normally

PR comming soon

**Comment by boelle** [comment]
done with header tags also

but I'm VERY unsure on the 2 files in the ht_mailchimp_360 subfolder

**Comment by boelle** [comment]
assuming that all PR's for PHP7 are ok what is the list of things we need to have a go at?

modules/content is one but what else?

**Comment by boelle** [comment]
Just checking stuff on my test subdomain... When i click security checks in admin i get this one:

Warning: Creating default object from empty value in ***/test/admin/security_checks.php on line 82

Fatal error: Uncaught Error: Call to undefined method stdClass::pass() in ***/test/admin/security_checks.php:87 Stack trace: #0 {main} thrown in ***/test/admin/security_checks.php on line 87

does this one depend on admin classes? that one i still miss to update but has it ready on desktop

**Comment by boelle** [comment]
Do we have anything other than modules/content to update?

I was after something light for the weekend as i will be on and off a lot

**Comment by boelle** [comment]
@gburton could you update the list in the first post? just so we know what is missing

and maybe expand what "other" covers

**Comment by boelle** [comment]
any news?

**Comment by BrockleyJohn** [comment]
@boelle 'other' is all the other directories not already listed above... in fact the only ones with code in are /ext and /install and I found nothing to change.

I guess I'll do modules / content myself so we can put this lot to bed.

**Comment by boelle** [comment]
i can help out with that... what if one of us start from the bottom and the other from the top?

**Comment by BrockleyJohn** [comment]
Sorry @boelle missed your comment - they're done now. I think at this point we've been across everything, so the next step is some end-to-end testing - I'll create another branch and merge all the requests into it so we've a consolidated set of the changes...

**Comment by boelle** [comment]
well i have pulled everything before today and have run it on a live shop without any issues

but if we branch with all changes in there i will pull that and run everything through a file compare tool so i dont miss any mods i have done

**Comment by BrockleyJohn** [comment]
Glad to hear you've been testing it all. All PR branches merged into this one on latest edge and conflicts resolved:
[https://github.com/BrockleyJohn/Responsive-osCommerce/tree/php7_consolidated](url)

**Comment by boelle** [comment]
will give it a go during the day or tomorrow, need to do a file compare 

the owner of the php7 driven site sudden said that he found another person and solution 

but i manage 2 other stores with the same setup... same hoster too, the servers for them have not been upgraded yet.... 

so will try and get this tested before the new guy steps in

**Comment by boelle** [comment]
Q... have been doing file compare for most of the day only interrupted by a dinner out

i came to includes\modules\action_recorder

i see     function ar_reset_password()  and not function __construct()

is that on purpose?

**Comment by BrockleyJohn** [comment]
Sorry - I assumed you were using github/sourcetree for the compare... when I first posted the link it was short of a couple of things that were added shortly afterwards; action recorder which I missed out and header tags which never had a pull request.

**Comment by boelle** [comment]
Oki, will pull a New copy then

What i'm doing is comparing my live shop with the consolidated branch so I
get as close to that I can and still have my mods

It will go live at a sub domain and be tested and then be moved to the main
site

Den 17/08/2016 09.39 skrev "John Ferguson" <notifications@github.com>:

**Comment by boelle** [comment]
did you also miss \includes\classes\order.php ?

**Comment by BrockleyJohn** [comment]
No, but I think the change must have been dropped in a merge conflict that wasn't flagged up. I have manually corrected.

**Comment by boelle** [comment]
oki.. i got started on admin section and will work my way through it

if i find anything i will leave a note.

**Comment by boelle** [comment]
think i got all files compared etc... now to a little error fixing i cant figure quite

`Warning: require(/includes/classes/action_recorder.php) [function.require]: failed to open stream: No such file or directory in /var/www/xn--pizzalg-v1a.dk/test/admin/includes/classes/action_recorder.php on line 13

Warning: require(/includes/classes/action_recorder.php) [function.require]: failed to open stream: No such file or directory in /var/www/xn--pizzalg-v1a.dk/test/admin/includes/classes/action_recorder.php on line 13

Fatal error: require() [function.require]: Failed opening required '/includes/classes/action_recorder.php' (include_path='.') in /var/www/xn--pizzalg-v1a.dk/test/admin/includes/classes/action_recorder.php on line 13`

line 13 look like this:   

`  require(DIR_FS_CATALOG . 'includes/classes/action_recorder.php');`

compared that one extra and they are the same on the server and the link you gave

cant it be this line in config:

  define('DIR_FS_CATALOG', dirname($_SERVER['SCRIPT_FILENAME']) . '/');

and if so could it simply be that my web hoster does not report the right path back?

will try with full path and see how it goes but i dont want to mod files more than absolute needed to still run the store as normal

**Comment by boelle** [comment]
that was the case... now it says that max login has been reached and that i should try again in 5 mins

even if i do create a new admin in mysql..

**Comment by boelle** [comment]
found a temp fix

/admin/login.php

arround line 39

if ($actionRecorder->canPerform()) {

i change it to

if ($actionRecorder->canPerform() || !tep_not_null($actionRecorder->_module)) {

now i can login but the box to the left is gone... so for now i can just stare at the pretty graphs

**Comment by boelle** [comment]
doohhh... missed one line in admin config :-/

going through everything in admin to catch anything odd... will try and keep that in one post

**Comment by boelle** [comment]
first one:

Parse error: syntax error, unexpected '_INVOICE' (T_STRING) in /var/www/xn--pizzalg-v1a.dk/test/admin/orders.php on line 119

line looks like:

            <td class="smallText" align="right"><?php echo tep_draw_button(IMAGE_ORDERS_INVOICE, 'document', tep_href_link('orders.php'_INVOICE, 'oID=' . $HTTP_GET_VARS['oID']), null, array('newwindow' => true)) . tep_draw_button(IMAGE_ORDERS_PACKINGSLIP, 'document', tep_href_link('orders.php'_PACKINGSLIP, 'oID=' . $HTTP_GET_VARS['oID']), null, array('newwindow' => true)) . tep_draw_button(IMAGE_BACK, 'triangle-1-w', tep_href_link('orders.php', tep_get_all_get_params(array('action')))); ?></td>



admin/sec_dir_permissions.php first char on each line is 1 short so ext becomes xt


that is what i can finde on the admin side... will do another and more slowly compare but i think i got it right....

could it be usefull if i fork the link above but then overwrite with my copy?

**Comment by boelle** [comment]
now going through settings and added navbar etc

http://**********/admin/modules.php?set=action_recorder&module=ar_contact_us&action=edit

Warning: reset() expects parameter 1 to be array, null given in /var/www/xn--pizzalg-v1a.dk/test/admin/modules.php on line 239

Warning: Variable passed to each() is not an array or object in /var/www/xn--pizzalg-v1a.dk/test/admin/modules.php on line 240

could this be an error somewhere on my part? and if so where should i look?

**Comment by BrockleyJohn** [comment]
I can reproduce the admin/orders error but not the action recorder one; the latter is in the code that produces the module edit panel, processing what should be an array of the module keys. If it only occurs on this module, it should be an error with the specific module but I would have expected earlier code to throw a not-an-array error.

**Comment by BrockleyJohn** [comment]
admin/orders error comes from master - PR raised with fix

**Comment by boelle** [comment]
i will try to see if i can figure the action recorder one

today will be spent doing languages... a lot have happend here too, but i assume that it will just be moving  lines arround in the files

**Comment by boelle** [comment]
Q... FILENAME_DEFAULT that should point to index.php right?

i had made a few custom info box items a year or so back that does not work when i click the continue button

**Comment by BrockleyJohn** [comment]
Yes. If you go back to a commit before filename constants were removed you can use includes/filenames.php as a crib sheet (look-up).

**Comment by boelle** [comment]
Implemented the fix...

now i get 

 Parse error: syntax error, unexpected '$order_query' (T_VARIABLE), expecting function (T_FUNCTION) in /var/www/*******/test/admin/includes/classes/order.php on line 27

**Comment by boelle** [comment]
gone through admin language and only need to translate a few.... once that is done i will go over admin again and check that everything works

if i find anything i will do them each a seperate issues as i think  it has less to do with aftermath of the php7 upgrade

**Comment by BrockleyJohn** [comment]
> Parse error: syntax error, unexpected '$order_query' (T_VARIABLE), expecting function (T_FUNCTION) in /var/www/*******/test/admin/includes/classes/order.php on line 2

@boelle you've either lost the line 
>     function query($order_id) {
gained an extra }, or something similar
The class in the github branch is fine and works in my test environment

**Comment by boelle** [comment]
it was the first thing,

and also i see that line 16 is set to 

function __construct($order_id) {

but on github its 


    function order($order_id) {

is both equal correct or should i compare all the classes?

**Comment by boelle** [comment]
\admin\includes\modules\security_check\ & \admin\includes\modules\newsletters\ & \admin\includes\modules\dashboard\ & \includes\modules\cfg_modules\ & \admin\includes\classes\

all modules in here are not updated to php7 in the way most other things are..

is this on purpose or have we just not come that far yet?

\admin\includes\classes\ only a few files in here can be exluded.

**Comment by BrockleyJohn** [comment]
@boelle ... oh yes, they are! At least in the branch of the relevant pull request and in the branch of the consolidated changes, eg [here](https://github.com/BrockleyJohn/Responsive-osCommerce/blob/php7_consolidated/admin/includes/modules/cfg_modules/cfgm_action_recorder.php)

I'm guessing it's something to do with the way you've built your test environment

**Comment by boelle** [comment]
my test setup is not complicated... i just pull down and ftp upload...

i guess those few golder must have slipped my eye somehow

but so far everything is going ok.... filling in the SEO fields now

tomorrow i will start to test it for real with a purchase while the store is closed.... nothing beats a live scenario... only difference for me is that i test from a sub domain..

**Comment by boelle** [comment]
could not sleep so why not check all files once more... check :+1: 
placed orders with various shipping methods and payment options :+1: 
even my danish payment gateway for credit cards works :+1:

**Comment by BrockleyJohn** [comment]
@gburton Gary, I think we're now at the end of this. Would you like a pull request from the consolidated branch? It will save you a bunch of work resolving all the conflicts with the global vars changes (at least one of which wasn't flagged), and it preserves all the commits and change authorship from the individual pull requests.

**Comment by boelle** [comment]
I agree....

tested it both on php5.6 and 7

**Comment by boelle** [comment]
@gburton when can we expect the hard work to be merged in?

its a shame that it should hang in limbo for this long.

**Comment by gburton** [comment]
When I am ready and have time.  Not before.

**Comment by boelle** [comment]
fair enough.... 

@BrockleyJohn will it be a pain for you to keep your consolidated branch up to date so we can use that one until the consolidated changes are merged in?

**Comment by BrockleyJohn** [comment]
Nope - but I'm in the process of rebuilding it without the dodgy commits - there will be a new pr for the new branch when it's ready for retesting

**Comment by boelle** [comment]
oki... i'm current until a few days ago

and no problems observed on the 2 live shops i have my  dirty hands on

will compare what i have and merge in the new changes and test them on a subdomain and if it goes well it will go live soon after

**Comment by boelle** [comment]
was todays close because that it was merged?

**Comment by BrockleyJohn** [comment]
@boelle  nope - but your work's not wasted
I'm keeping the later consolidated branch up to date until core is php7-ready
atm I'm in the middle of the ppApp - there are a lot of class constructors!
[status here](https://github.com/BrockleyJohn/Responsive-osCommerce/wiki/PHP-7-changes)

**Comment by mcmannehan** [comment]
Hi together, i am new in Github, i work with osCommerce since the start of osCommerce :-)
I did start to change every class file to PHP7 compatibility.  I change every thing to __construct

i don't know now how far you are with the devlopment. 

I changed in /admin/security_checks.php at line 79
//$secCheck = $module['class'];
$secCheck = ${$module['class']};

I changed in the /admin/categories.php at line 432
//$pInfo->objectInfo($product);
$pInfo->__construct($product);

I changed in the /admin/banner_manager.php at line 230
  //$bInfo->objectInfo($banner);
   $bInfo->__construct($banner);
} elseif (tep_not_null($HTTP_POST_VARS)) {
   //$bInfo->objectInfo($HTTP_POST_VARS);
   $bInfo->__construct($HTTP_POST_VARS);
 }

I also changed some MySQL Querry for compatibillty of MySQL 5.7 mode.

now (only the admin) everything is working with PHP7.0.14 and MySql 5.7.

**Comment by BrockleyJohn** [comment]
@mcmannehan Had you followed the link in the comment immediately before yours, you would have saved yourself a lot of work!

All the php7 changes are already done and tested - available in their own branch (and capable of pulling into master at the drop of a hat). When changes take place in master and merge conflicts arise, I recreate the php7 branch against the latest changes.

**Comment by gburton** [comment]
@BrockleyJohn  John your effort is not in vain TY.  I am awaiting the 2.3.5 changes (which may have php7 changes already as part of it).  If not, yours goes stright into Core.

**Comment by mcmannehan** [comment]
@BrockleyJohn Thank you for the info, but i did download https://github.com/gburton/Responsive-osCommerce and there was no changes in the Class files. All other links here in the comments give me an 404 error. Maybe I overlooked something. Now i did start my own project here in Github, because i do MySql query changes for MySQL 5,7.

https://github.com/mcmannehan/osCommerce-2.3.4-Responsiv

**Comment by BrockleyJohn** [comment]
For the avoidance of doubt, this is the link to the page on the wiki in my fork which I keep updated with the latest branch: [https://github.com/BrockleyJohn/Responsive-osCommerce/wiki/PHP-7-changes](https://github.com/BrockleyJohn/Responsive-osCommerce/wiki/PHP-7-changes)

**Comment by mcmannehan** [comment]
@BrockleyJohn Thanks again i saw it already.

**Comment by mcmannehan** [comment]
@BrockleyJohn 
your repositories https://github.com/BrockleyJohn/Responsive-osCommerce
is missing __construct in /admin/includes/modules/dashboard/d_paypal_app.php

**Comment by BrockleyJohn** [comment]
@mcmannehan thanks, Manfred. That's now updated. Must have been testing with my eyes shut!

**Comment by mcmannehan** [comment]
@BrockleyJohn  your welcome
and btw i use your respositorie for my stand alone pimp project.

This are the only canges for MySQL 5.7  suitable

Changes in
3065 - Expression #1 of ORDER BY clause is not in SELECT list, references column 'xxx.p.products_ordered' which is not in SELECT list; this is incompatible with DISTINCT
/catalog/includes/modules/boxes/bm_best_sellers.php >>> Mysql query

1055 - Expression #1 of ORDER BY clause is not in GROUP BY clause and contains nonaggregated column 'web228_4.o.date_purchased' which is not functionally dependent on columns in GROUP BY clause; this is incompatible with sql_mode=only_full_group_by
/catalog//includes/modules/also_purchased_products.php >>> Mysql query

3065 - Expression #1 of ORDER BY clause is not in SELECT list, references column 'web228_4.p.products_date_added' which is not in SELECT list; this is incompatible with DISTINCT
/catalog//includes/modules/content/index_nested/cm_in_new_products.php >>> Mysql query
/catalog//includes/modules/content/index/cm_i_new_products.php >>> Mysql query
/catalog//includes/modules/new_products.php >>> Mysql query

**Comment by boelle** [comment]
@BrockleyJohn

could the following error be something missing for php7? or is it more my fault?

Warning: Creating default object from empty value in /var/www/xn--pizzalg-v1a.dk/public_html/admin/security_checks.php on line 82

Fatal error: Uncaught Error: Call to undefined method stdClass::pass() in /var/www/xn--pizzalg-v1a.dk/public_html/admin/security_checks.php:87 Stack trace: #0 {main} thrown in /var/www/xn--pizzalg-v1a.dk/public_html/admin/security_checks.php on line 87

**Comment by mcmannehan** [comment]
@boelle what rep you use?
 @BrockleyJohn  i found a missing changes again in /admin/includes/modules/dashboard/d_security_checks.php 
Not change to __construct

This rep https://github.com/BrockleyJohn/Responsive-osCommerce

**Comment by mcmannehan** [comment]
@boelle look here the file must look like that:

https://github.com/mcmannehan/osCommerce-WDW-PV7.0-Pimp-Version-Responsiv/blob/master/catalog/admin/includes/modules/dashboard/d_security_checks.php

**Comment by boelle** [comment]
i use https://github.com/BrockleyJohn/Responsive-osCommerce/tree/php7_compatibility_01

**Comment by mcmannehan** [comment]
@boelle my pimp version is based on that. may by you did something wrong... 
With all do respect let say with your words: your fault...

**Comment by boelle** [comment]
and how is this my fault? i dont not change the file

maybe we should stick the more clean version and not any pimp'ed ones

**Comment by mcmannehan** [comment]
@boelle i don't know what you did, 
but this rep
https://github.com/BrockleyJohn/Responsive-osCommerce/tree/php7_compatibility_01
and this rep
https://github.com/mcmannehan/osCommerce-WDW-PV7.0-Pimp-Version-Responsiv
is running without this error.

May be copy all files from the admin folder from
https://github.com/BrockleyJohn/Responsive-osCommerce/tree/php7_compatibility_01
again. i did this on my local server today and nothing happens.

**Comment by BrockleyJohn** [comment]
@boelle check if you've got any extra files in admin/includes/modules/security_check/

It looks like you're loading something that's not a valid class definition

**Comment by boelle** [comment]
@BrockleyJohn just downloaded your php7 repro and compared

they are the same...

will run everything through a file diff tool to make sure all files are up to date. cant hurt to be sure that they are

**Comment by BrockleyJohn** [comment]
Weird. Actually a dodgy extra file gives a different error. Are you sure that line 79 of security_checks.php is
`    $secCheck = ${$module['class']};
`
with the curly braces

**Comment by boelle** [comment]
upps.... that was the one :-D

actually i raised an issue in your repro asking if it was possible to get some notification when you update so we can keep up

**Comment by boelle** [comment]
i was more thinking if @gburton made minor changes and you had to rebase your php7 repro to keep it current

**Comment by mcmannehan** [comment]
@BrockleyJohn 

> Weird. Actually a dodgy extra file gives a different error. Are you sure that line 79 of security_checks.php is
$secCheck = ${$module['class']};
with the curly braces

I did wrote something about that on stackoverflow.com but didn't found again, no time to search.
If someone have the time. ?!?

**Comment by BrockleyJohn** [comment]
I had forgotten the exact error... discussed at length here: [http://forums.oscommerce.com/topic/409316-php7-dynamic-classarray-handling-gotcha/](http://forums.oscommerce.com/topic/409316-php7-dynamic-classarray-handling-gotcha/)

**Comment by mcmannehan** [comment]
today i start wondering:
i go here https://github.com/BrockleyJohn/Responsive-osCommerce/tree/php7_compatibility_01
and click let say /admin/includes/classes/action_recorder.php all changes made for PHP7.

i did open a subdomain https:/brockley-john.webdesign-wedel.de my Provider use Plesk very i can load direct this https://github.com/BrockleyJohn/Responsive-osCommerce.git coz this git get if you
here https://github.com/BrockleyJohn/Responsive-osCommerce/tree/php7_compatibility_01

But if i open the files on my server;
let say /admin/includes/classes/action_recorder.php no changes for PHP7.

Did i miss somethimg? Why i see online https://github.com/BrockleyJohn/Responsive-osCommerce/tree/php7_compatibility_01 the PHP7.x changes?

They give me this git for download https://github.com/BrockleyJohn/Responsive-osCommerce.git
but the files inside don't have the PHP changes

**Comment by BrockleyJohn** [comment]
You're loading the master branch. You need to find out how to load a project branch.

**Comment by mcmannehan** [comment]
@BrockleyJohn 
The evaluation order in PHP7 is left-to-right instead of right-to-left so you have to make the order explicit in the statement:

$secCheck = ${$module['class']};

This makes sure PHP5 and PHP7 use the same order.

Yes, this  is it. Thx John!

**Comment by mcmannehan** [comment]
@BrockleyJohn 
About to get the right version: I download Zip file: https://github.com/BrockleyJohn/Responsive-osCommerce/archive/php7_compatibility_01.zip
This version will be online soon for compare. It will run with the follow Server conditions:
PHP7.0.15, MySQL 5.7, Apache 2.2

**Comment by mcmannehan** [comment]
@BrockleyJohn 
at https://brockley-john.webdesign-wedel.de/ runs https://github.com/BrockleyJohn/Responsive-osCommerce/tree/php7_compatibility_01

**Comment by mcmannehan** [comment]
@BrockleyJohn 
at https://osc.webdesign-wedel.de/ runs https://github.com/mcmannehan/osCommerce-WDW-PV7.0-Pimp-Version-Responsiv

**Comment by mcmannehan** [comment]
@BrockleyJohn 
I recommand the changes you can see at https://brockley-john.webdesign-wedel.de

**Comment by mcmannehan** [comment]
@gburton 

newste version installed PHP7.1.2 MySQL 5.7 all went good, but
install/install.php?step=4 appears with

Deprecated: Methods with the same name as their class will not be constructors in a future version of PHP; PasswordHash has a deprecated constructor in /includes/classes/passwordhash.php on line 32

**Comment by BrockleyJohn** [comment]
@mcmannehan you should be taking from PR #533 and testing that one. Thanks!

**Comment by mcmannehan** [comment]
@gburton okay thanks
